File: daemon_single.pl
#!/usr/bin/perl -w

my $file = "/var/log/httpd/access_log";


sub mail_out {
    my ($missing_link, $referrer) = @_;

    print "
    Write your code to send out the mail alerts
    $missing_link is missing refered from $referrer
    ";

}


if (! fork()) {
    $0 = "log_reader";

    print ("open file\n");
    open(LOG, "<$file") or die("Unable to open ");

    print ("Seek to the end of file\n");
    seek(LOG, 0, 2); #seek to the end of the file.... listen to only new entries

    print ("loop\n");
    while(1) {
        while(<LOG>) {
            if (/GET\s{1,}(\S{1,})\s{1,}\S{1,}\s{1,}404\s{1,}\d{1,}\s{1,}(\S{1,})/) {
                mail_out($1, $2)
            }
        }
        sleep(10);        
    }

} else {
    print ("Daemonizinng...\n");
    exit(0);
}

File: daemon_single.py
#!/usr/bin/env python3
import os
import sys
import time
import re

LOG_FILE = "/var/log/httpd/access_log"

def mail_out(broken_link, referrer):
    print (
    """
    Write your code to send out the mail alerts
    {missing_link} is missing refered from {referrer}
    """.format(missing_link=broken_link, referrer=referrer)
    )


def main():
    pid = os.fork()

    if(pid):
        print("Daemonizing...\n")
        sys.exit(0)
    else:
        try:
            with open(LOG_FILE) as log:
                log.seek(0, 2) #seek to the end
                while True:
                    for line in log:
                        match = re.search(
                            r'GET\s{1,}([^\s]{1,})\s{1,}[^\s]{1,}\s{1,}404\s{1,}\d{1,}\s{1,}([^\s]{1,})', 
                            line
                        )
                        if match:
                            mail_out(match.group(1), match.group(2))
                    time.sleep(1)
        except OSError:
            print ("Unable to open file {}\n".format(LOG_FILE))
            sys.exit(1)

if __name__ == "__main__":
    main()
